ROX-25909: CRS Support for roxctl

mclasmeier opened the pull request:
### Description

This PR implements CRS management capabilities (generating, listing, revoking) in roxctl.
These features are currently hidden behind a feature flag -- **not yet exposed to users**.

**NOTE**: A follow-up commit with some sort of unit-testing (bats, probably) needs to be added.

### User-facing documentation

- [x] CHANGELOG update is not needed (until the feature is actually exposed)
- [x] documentation changes not needed (until the feature is actually exposed)

### Testing and quality

- [x] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [x] added unit tests

#### How I validated my change

Currently manual testing.
CI test added.

mclasmeier commented:
> #### ️✅ There are no secrets present in this pull request anymore.
> If these secrets were true positive and are still valid, we highly recommend you to revoke them. Once a secret has been leaked into a git repository, you should consider it compromised, even if it was deleted immediately. Find [here](https://docs.gitguardian.com/platform/remediate/remediate-incidents) more information about risks.
> 
> 🦉 [GitGuardian](https://dashboard.gitguardian.com/auth/login/?utm_medium=checkruns&utm_source=github&utm_campaign=cr1) detects secrets in your source code to help developers and security teams secure the modern development process. You are seeing this because you or someone else with access to this repository has authorized GitGuardian to scan your pull request.

This is about test credentials.

mtodor commented on roxctl/central/crs/list.go:
```diff
@@ -0,0 +1,72 @@
+	defer utils.IgnoreError(conn.Close)
+	svc := v1.NewClusterInitServiceClient(conn)
+
+	tabWriter := tabwriter.NewWriter(cliEnvironment.InputOutput().Out(), 4, 8, 2, '\t', 0)
```
nitpick: we can move this just before the first `fmt.Fprintln(tabWriter`.

mclasmeier replied:
done

mtodor reviewed (commented):
Nice work!

It would be good to have tests for this functionality. In some cases, we have unit tests. In commands where we call server, we also have `bat` tests. It would also be good to cover it with an enabled/disabled flag.

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+	defer utils.IgnoreError(conn.Close)
+	svc := v1.NewClusterInitServiceClient(conn)
+
+	defer func() {
```
Did that in commit 29695307c1.

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+	if outFilename != "" {
+		outFile, err = os.OpenFile(outFilename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+		if err != nil {
+			return errors.Wrap(err, "opening output file for writing CRS")
```
done

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+	crs := resp.GetCrs()
+	meta := resp.GetMeta()
+
+	cliEnvironment.Logger().InfofLn(`Successfully generated new CRS.
```
done.

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+			if outputFile == "" {
+				return common.ErrInvalidCommandOption.New("No output files specified with --output (for stdout, specify '-')")
+			}
+			if outputFile != "" {
+				if outputFile == "-" {
```
of course.

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+	}
+	_, err = outWriter.Write(crs)
+	if err != nil {
+		return errors.Wrapf(err, "writing CRS to %s", stringutils.FirstNonEmpty(outFilename, "<stdout>"))
```
Fine to reconsider this for a follow-up PR?
(The CRS topic is far from done)

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+	c := &cobra.Command{
+		Use:   "generate <crs name>",
+		Short: "Generate a new CRS",
+		Long:  "Generate a new CRS for bootstrapping a new StackRox secured cluster",
```
done in commit b9b342a95ce01fca2060709900ee53a5c98b49e4

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+		RunE: func(cmd *cobra.Command, args []string) error {
+			name := args[0]
+			if outputFile == "" {
+				return common.ErrInvalidCommandOption.New("No output files specified with --output (for stdout, specify '-')")
```
thanks

mclasmeier commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,122 @@
+
+`,
+		meta.GetName(),
+		meta.GetCreatedAt().AsTime().Format(time.RFC3339),
```
I have the impression that `Format(time.RFC3339)` is currently used throughout the codebase, hence I would suggest to stick with this -- at least for the time being.

mclasmeier commented on roxctl/central/crs/revoke.go:
```diff
@@ -0,0 +1,94 @@
+	svc := v1.NewClusterInitServiceClient(conn)
+
+	idsOrNamesSet := set.NewStringSet(idsOrNames...)
+	if err = applyRevokeCRSs(ctx, cliEnvironment, svc, idsOrNamesSet); err != nil {
+		return err
+	}
+	return nil
```
changed

mclasmeier commented on pkg/features/list.go:
```diff
@@ -131,4 +131,7 @@ var (
 	NvdCvssUI = registerFeature("Display NVD CVSS score in UI", "ROX_NVD_CVSS_UI")
+
+	// ClusterRegistrationSecrets enables support for Cluster Registration Secrets (CRS), the next-gen init-bundles.
+	ClusterRegistrationSecrets = registerFeature("Enable support for Cluster Registration Secrets (CRS)", "ROX_CLUSTER_REGISTRATION_SECRETS")
```
The backend changes so far did not require any feature flags, this is the first one.

mclasmeier commented on roxctl/central/crs/revoke.go:
```diff
@@ -0,0 +1,94 @@
+	}
+	printResponseResult(cliEnvironment.Logger(), revokeResp)
+
+	if len(revokeResp.GetCrsRevocationErrors()) == 0 {
+		cliEnvironment.Logger().InfofLn("Revoked %d CRS(s)", len(revokeIds))
+	} else {
+		cliEnvironment.Logger().ErrfLn("Failed. Revoked %d of %d CRS(s)", len(revokeResp.GetRevokedIds()), len(revokeIds))
+	}
```
good idea, done.

mclasmeier commented:
@mtodor Thank you for the review. I hope I have addresses all your feedback.
I also did a rebase from master, because of `roxvet` problems I've been hitting.

mclasmeier commented on roxctl/central/crs/common.go:
```diff
@@ -0,0 +1,26 @@
+	attributePrecedence := []string{"email", "name", "userid"}
+	attributes := make(map[string]string)
+
+	if user == nil {
```
done

mclasmeier commented:
@mtodor just added a unit test.

mclasmeier commented:
/retest

mclasmeier commented:
/retest

mtodor commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,108 @@
+
+	req := v1.CRSGenRequest{Name: name}
+	resp, err := svc.GenerateCRS(ctx, &req)
+	if err != nil {
```
nitpick (just in case you want to go a step further): it would be great if we could distinguish between generation failed and central does not support that functionality. And provide better feedback to users.

mclasmeier replied:
done

mtodor commented on roxctl/central/crs/revoke.go:
```diff
@@ -0,0 +1,92 @@
+
+func printResponseResult(logger logger.Logger, revokeIds []string, resp *v1.CRSRevokeResponse) {
+	for _, id := range resp.GetRevokedIds() {
+		logger.InfofLn("Revoked %q", id)
```
nitpick: I'm wondering if we could improve the information here and avoid confusion. Because a user can provide a CRS name, and the user will see only the ID of the removed CRS. It's not clear what happened there.

If we change `var revokeIds []string` - to `var revokeIds map[string]string` and store `id->name`, we could improve the log here and provide both information.

mclasmeier replied:
done

mtodor commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,108 @@
+	c := &cobra.Command{
+		Use:   "generate <CRS name>",
+		Short: "Generate a new Cluster Registration Secret",
+		Long:  "Generate a new CRS for bootstrapping a new Secured Cluster",
```
nitpick: For other commands, we use a full name, `Cluster Registration Secret (CRS)` - it is not so important if it is an abbreviation or full, but let's keep it consistent across all commands.

mtodor reviewed (approved):
It looks good. You made nice improvements.

I have added a few nitpicks (nothing crucial) and a suggestion to extend testing to errors and not only "happy paths".

mclasmeier commented on tests/roxctl/bats-tests/cluster/crs-lifecycle.bats:
```diff
@@ -0,0 +1,39 @@
+    fi
+}
+
+@test "CRS can be issued" {
```
I have added two tests for failure situations. The second point is not easy to test right now (in a meaningful way, I mean). But this will change once I have changed the backend CRS issuing implementation so that, e.g. name duplicates are disallowed.

mclasmeier commented:
@mtodor Thank you again for the feedback. I hope I have addressed everything.

parametalol commented on roxctl/central/crs/generate.go:
```diff
@@ -0,0 +1,113 @@
+	c := &cobra.Command{
+		Use:   "generate <CRS name>",
+		Short: "Generate a new Cluster Registration Secret",
+		Long:  "Generate a new Cluster Registration Secret (CRS)  for bootstrapping a new Secured Cluster",
```
```suggestion
		Long:  "Generate a new Cluster Registration Secret (CRS) for bootstrapping a new Secured Cluster",
```
